op-node: Implement fjord RLP & Channel Bank Size limit Increases

trianglesphere opened the pull request:
**Description**

Implements https://github.com/ethereum-optimism/specs/pull/165 & https://github.com/ethereum-optimism/specs/pull/170 in the op-node.

I made several changes to span batch parsing which I believe to be spec compliant.
* The span batch size is no longer checked inside the span batch parsing code. The limit is enforced at the RLP reader level.
* The buffers allocated by bitfields are no longer checked for size. When looking at the code, we always pass in a size limited field as the bitlength (specifically a field which is limited to `MAX_SPAN_BATCH_ELEMENT_COUNT`). That means that the largest buffer possible to allocate is 1.25 million bytes while the code was checking against a size limit of 10 million bytes.

**Tests**

I added chain spec tests. No tests inside the channel bank / channel reader; however I made the old constant private so they can only be accessed through these function.

**Metadata**

- Fixes https://github.com/ethereum-optimism/protocol-quest/issues/246

sebastianst commented on op-node/rollup/derive/params.go:
```diff
@@ -30,14 +30,22 @@ const DerivationVersion1 = plasma.TxDataVersion1
 // MaxChannelBankSize is the amount of memory space, in number of bytes,
-// till the bank is pruned by removing channels,
-// starting with the oldest channel.
+// till the bank is pruned by removing channels, starting with the oldest channel.
 const MaxChannelBankSize = 100_000_000
 
+// MaxChannelBankSizeV1 is the amount of memory space, in number of bytes,
+// till the bank is pruned by removing channels, starting with the oldest channel.
+// This limit is applied after Fjord
+const MaxChannelBankSizeV1 = 1_000_000_000
+
 // MaxRLPBytesPerChannel is the maximum amount of bytes that will be read from
 // a channel. This limit is set when decoding the RLP.
 const MaxRLPBytesPerChannel = 10_000_000
 
+// MaxRLPBytesPerChannelV1 is the maximum amount of bytes that will be read from
+// a channel after Fjord. This limit is set when decoding the RLP.
+const MaxRLPBytesPerChannelV1 = 100_000_000
```
Maybe we should remove these constants from the public API of the `derive` package to avoid accidentally referencing the old constant. We would instead force ppl to get the protocol parameters from a rollup config (see other comment), avoiding using a go `const` that isn't a const across forks any more.

We could then also rename the private constants with suffixes `V0` and `V1` or even `...Bedrock` and `...Fjord`.

sebastianst commented on op-node/rollup/derive/channel_bank.go:
```diff
@@ -60,14 +60,21 @@ func (cb *ChannelBank) Origin() eth.L1BlockRef {
 	return cb.prev.Origin()
 }
 
+func (cb *ChannelBank) maxSize() int {
+	if cb.Config.IsFjord(cb.Origin().Time) {
+		return MaxChannelBankSizeV1
+	}
+	return MaxChannelBankSize
+}
```
Wdyt about starting to treat the `rollup.Config` more like a specs object and add this logic instead to it? Then we have the protocol rules encapsulated with the rollup config, not as methods on various structs.

```go
func (c *Config) MaxChannelBankSize(t uint64) int {
	if c.IsFjord(t) {
		return maxChannelBankSizeV1 // or ...Ecotone
	}
	return maxChannelBankSizeV0 // or ...Bedrock
}
```

sebastianst commented on op-node/rollup/derive/channel_in_reader.go:
```diff
@@ -44,9 +44,16 @@ func (cr *ChannelInReader) Origin() eth.L1BlockRef {
 	return cr.prev.Origin()
 }
 
+func (cr *ChannelInReader) maxRLPBytePerChannel() int {
+	if cr.cfg.IsFjord(cr.prev.Origin().Time) {
+		return MaxRLPBytePerChannelV1
+	}
+	return MaxRLPBytePerChannel
+}
```
Same, maybe use `func (c *Config) MaxRLPBytePerChannel(t uint64) int` instead?

sebastianst commented on op-node/rollup/derive/params.go:
```diff
@@ -30,14 +30,22 @@ const DerivationVersion1 = plasma.TxDataVersion1
+
 // MaxRLPBytesPerChannel is the maximum amount of bytes that will be read from
 // a channel. This limit is set when decoding the RLP.
 const MaxRLPBytesPerChannel = 10_000_000
```
There are more references to this const, e.g. in span channel out and `op-e2e/actions/garbage_channel_out.go`. We may want to replace all references.

trianglesphere committed:
op-node: Increase MaxChannelBankSize with Fjord

This also creates a ChainSpec object which is responsible for returning protocol
parameters. We use a different object than the rollup.Config because the config
is primarily a disk representation & does not concern itself with protocol constants.

trianglesphere committed:
op-node: Increase MaxRLPBytesPerChannel with Fjord

sebastianst commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -204,9 +207,6 @@ func (bp *spanBatchPayload) decodePayload(r *bytes.Reader) error {
 
 // decode reads the byte encoding of SpanBatch from Reader stream
 func (b *RawSpanBatch) decode(r *bytes.Reader) error {
-	if r.Len() > MaxSpanBatchSize {
-		return ErrTooBigSpanBatchSize
-	}
```
So this is now handled by individual `decode...` functions testing against `MaxSpanBatchElementCount` right?

trianglesphere replied:
This specific limit is checked by the channel_in_reader

sebastianst commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -58,6 +58,9 @@ func (b *RawSpanBatch) GetBatchType() int {
 
 // decodeOriginBits parses data into bp.originBits
 func (bp *spanBatchPayload) decodeOriginBits(r *bytes.Reader) error {
+	if bp.blockCount > MaxSpanBatchElementCount {
+		return ErrTooBigSpanBatchSize
+	}
```
is this (and other limits) tested?

trianglesphere replied:
I only added these when these limits where being tested. IMO the tests for these limits are not great because it is modifying the internal state of `*spanBatchPayload` to be invalid and then calling this function.

sebastianst commented on op-node/rollup/derive/span_batch_txs.go:
Are the new limits in this file tested?

If we add `MaxSpanBatchElementCount` to the spec, it might get easier to test because we could set low limits for tests. Where we consume the spec, we could also only consume an interface of the value getters that we need to set low limits during tests.

sebastianst replied:
Dealing with this in https://github.com/ethereum-optimism/client-pod/issues/850

sebastianst commented on op-node/rollup/derive/span_batch_util.go:
why are the checks safe to remove in this file?

trianglesphere replied:
`bitlength` is capped to `MAX_SPAN_BATCH_ELEMENT_COUNT` == 10,000,000. 8 bits per bit => bufLen is capped at 1,250,000 < MaxSpanBatchSize (10,000,000)

sebastianst reviewed (approved):
LGTM, I like the introduction of a spec. 

I have a few open questions that might lead to changes.

If we consume only an interface of the spec (or even only the value getters from the spec that we need), it might become easier in tests to set low values, as the spec can then be mocked.
